### PR TITLE
Make all nightly and nightly_protocol flags consistent.

### PR DESCRIPTION
**Pull request by robin-near**
Made a script to do this, see scripts/fix_nightly_feature_flags.py.
 1. If a crate declares the nightly feature, it should also declare nightly_protocol, and vice versa.
 2. nightly_protocol should not activate any other feature, only nightly_protocol features of dependency crates.
 3. nightly should activate nightly_protocol of the same crate.
 4. If crate A depends on crate B and crate B depends on crate C (both transitively), and A and C both declares a nightly feature, then B must also declare a nightly feature.
 5. If crate A directly depends on crate B and they both declare a nightly feature, then A's nightly feature must activate B's nightly feature; same with nightly_protocol.
 6. If crate A depends transitively on crate B, they both declare a nightly feature, and there's a feature fB declared by B that is activated by B's nightly feature, and there's also a feature fA declared by A with the same name as fB, then A's nightly feature must also activate fA.

And cleaned up existing nightly feature declarations.

### CODE REVIEW

**Comment by nagisa** [comment]
I think this should (also) be a test/CI check. Having a script to run won't do much good if nobody bothers to run it and keep the state consistent.

File: scripts/fix_nightly_feature_flags.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+    def __init__(self, dir):
+        self.dir = dir
+        self.height = None
+        self.load_toml()
+
+    def load_toml(self):
+        self.toml = toml.load(self.dir + "/Cargo.toml")
+        self.name = self.toml["package"]["name"]
+        self.features = self.toml["features"] if 'features' in self.toml else {}
```
Comment (jakmeier) [inline scripts/fix_nightly_feature_flags.py:22]: (optional suggestion)
```suggestion
        self.features = self.toml.get("features", {})
```

File: scripts/fix_nightly_feature_flags.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+            for dep in crate.deps:
+                if dep.name not in visited:
+                    visited.add(dep.name)
+                    helper(dep)
+
+        for dep in self.deps:
+            if dep.name not in visited:
+                visited.add(dep.name)
+                helper(dep)
+        self.transitive_deps = result
```
Comment (jakmeier) [inline scripts/fix_nightly_feature_flags.py:87]: I think there shouldn't be any duplicate dependencies, right?

```suggestion
        for dep in self.deps:
            visited.add(dep.name)
            helper(dep)
        self.transitive_deps = result
```

File: scripts/fix_nightly_feature_flags.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+            crates_by_name[dep]
+            for dep in self.toml['dependencies'].keys()
+            if dep in crates_by_name
+        ]
+
+    def build_transitive_deps(self):
+        visited = set()
+        result = []
+
+        def helper(crate):
```
Comment (jakmeier) [inline scripts/fix_nightly_feature_flags.py:76]: nit: can we use a more descriptive name? Idk, maybe `visit_recursively` or `recursion_helper` or something like that.

**Review by jakmeier — APPROVED** [review]
Again, this script is awesome and will help many of us in the future! I remember in particular external contributors struggling with this as well when implementing their NEP.

Left some tiny style nits, all optional and can be ignored

File: scripts/fix_nightly_feature_flags.py (hunk 2 above)
Comment (robin-near) [reply scripts/fix_nightly_feature_flags.py:87]: I also thought this was not needed, but it turns out to be needed. I, uh, can't remember exactly what, but I spent at least an hour debugging this and there was some reason.

Comment (jakmeier) [reply scripts/fix_nightly_feature_flags.py]: ok, just keep it in then, it certainly doesn't break correctness to have it in :)

**Pull request merged**